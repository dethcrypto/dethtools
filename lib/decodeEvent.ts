import { EventFragment, Interface, Result } from '@ethersproject/abi'

export function decodeEvent(iface: Interface, eventProps: EventProps): DecodeEventResult | undefined {
  const { data, topics } = eventProps
  const events = iface.events

  let decodedTopics: Result | Record<string, unknown> | undefined
  let eventFragment: [string, EventFragment] | undefined

  for (const event of Object.entries(events)) {
    const name = event[1].name
    try {
      decodedTopics = iface.decodeEventLog(name, data, topics)
      eventFragment = event
    } catch (e) {}
  }
  if (decodedTopics && eventFragment) {
    let isNamedResult = false
    if (doContainsNamedKeys(decodedTopics)) {
      isNamedResult = true
      decodedTopics = filterNonNamedKeys(decodedTopics)
    }
    return { decodedTopics, eventFragment, isNamedResult }
  }
}

export interface EventProps {
  topics: readonly string[]
  data: string
}

export interface DecodeEventResult {
  decodedTopics: Result | Record<string, unknown>
  eventFragment: [string, EventFragment]
  isNamedResult: boolean
}

// @internal
export function filterNonNamedKeys<T extends object>(object: T): Record<string, unknown> | T {
  const obj: Record<string, unknown> = {}
  for (const [key, val] of Object.entries(object)) {
    if (isNaN(parseInt(key))) {
      obj[key] = val
    }
  }
  return obj
}

// @internal
export function doContainsNamedKeys<T extends object>(object: T): boolean {
  return Object.keys(object).some((key) => isNaN(parseInt(key)))
}
