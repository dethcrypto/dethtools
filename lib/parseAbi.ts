import { Interface } from '@ethersproject/abi'

export function parseAbi(rawAbi: string, defaultKeyword: string = 'function'): Interface | Error {
  const formated: string[] = []

  // check for JSON format
  try {
    return new Interface(rawAbi)
  } catch (e) {
    if (!(e instanceof Error && e.message.startsWith('Unexpected token'))) {
      return e as Error
    }
  }

  const parsed = rawAbi
    .split("',")
    .map((line) =>
      line
        .trim()
        .split('')
        .filter((char) => char !== "'")
        .join(''),
    )
    .filter((line) => line)

  for (const frag of parsed) {
    if (!findKeyword(frag)) {
      // if the fragment contains indexed, it's a event
      if (frag.includes('indexed')) {
        formated.push(`event ${frag}`)
      } else {
        formated.push(`${defaultKeyword} ${frag}`)
      }
    } else {
      formated.push(frag)
    }
  }

  return new Interface(formated)
}

const keywords = ['function', 'modifier', 'event', 'error', 'constructor', 'fallback', 'receive']

// @internal
export function lastOne<T>(array: T[]): T {
  return array[array.length - 1]
}

// @internal
export function findKeyword(frag: string): boolean {
  frag = frag.trim()

  let word = ''
  const match = new RegExp(/[a-zA-Z]/)

  for (const letter of frag.split('')) {
    if (!match.test(letter)) break
    else word += letter
  }

  for (const keyword of keywords) {
    if (keyword === word) return true
  }
  return false
}
