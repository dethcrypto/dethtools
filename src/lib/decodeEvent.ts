import {
  EventFragment,
  FormatTypes,
  Interface,
  ParamType,
} from '@ethersproject/abi';
import { version } from '@ethersproject/abi/lib/_version';
import { BigNumber } from '@ethersproject/bignumber';
import { Logger } from '@ethersproject/logger';

import { combinations as generateCombinations } from '../misc/combinations';

const logger = new Logger(version);

export interface EventProps {
  topics: string[];
  data: string;
}

interface AbstractDecodedEventResult {
  eventFragment: EventFragment;
  hasArgumentNames: boolean;
  signature: string;
  fullSignature: string;
}

export interface DecodedEventResultWithArgNames
  extends AbstractDecodedEventResult {
  hasArgumentNames: true;
  args: Record<string, string | BigNumber>;
}

export interface DecodedEventResultWithoutArgNames
  extends AbstractDecodedEventResult {
  hasArgumentNames: false;
  args: ReadonlyArray<string | BigNumber>;
}

export type DecodedEventResult =
  | DecodedEventResultWithArgNames
  | DecodedEventResultWithoutArgNames;

export function decodeEvent(
  iface: Interface,
  eventProps: EventProps,
): DecodedEventResult {
  try {
    const decoded = iface.parseLog(eventProps);
    /**
     * Ethers under the hood converts EventFragment to object-like form without format function, so we rehydrate it here.
     */
    const eventFragment = EventFragment.fromObject({
      ...decoded.eventFragment,
      _isFragment: false, // Otherwise this function noops
    });
    const hasArgumentNames = doesContainNamedKeys(decoded.args);
    const decodedEventResult = {
      args: hasArgumentNames ? omitNonNamedKeys(decoded.args) : decoded.args,
      eventFragment: eventFragment,
      hasArgumentNames,
      signature: decoded.signature,
      fullSignature: eventFragment.format(FormatTypes.full),
    };
    return decodedEventResult as DecodedEventResult;
  } catch (e: any) {
    switch (e.reason) {
      case 'no matching event':
        throw e;
      case 'data out-of-bounds':
        // At this point we know that we have a matching topic, but the number of indexed args is incorrect
        return tryToDecodeEvent(iface, eventProps);
      default:
        throw e;
    }
  }
}

/**
 * We assume that we don't have partial indexing information.
 * This may not be true, but it's an edge case and it's PITA to handle.
 */
export function tryToDecodeEvent(
  iface: Interface,
  eventProps: EventProps,
): DecodedEventResult {
  const indexedArgsCount = eventProps.topics.length - 1;
  const topichash = eventProps.topics[0];

  for (const name in iface.events) {
    if (topichash === iface.getEventTopic(name)) {
      const event = iface.getEvent(name);
      const inputsCombinations = combinationsMap(
        event.inputs,
        indexedArgsCount,
        (input: ParamType): ChangedParamType => ({
          arrayChildren: input.arrayChildren,
          components: input.components,
          name: input.name,
          type: input.type,
          arrayLength: input.arrayLength,
          baseType: input.baseType,
          indexed: true,
        }),
      );

      let error: unknown;
      for (const inputs of inputsCombinations) {
        const newEvent = EventFragment.fromObject({
          inputs,
          anonymous: event.anonymous,
          type: event.type,
          name: event.name,
        });

        iface.events[name] = newEvent;

        try {
          const args = iface.decodeEventLog(
            newEvent,
            eventProps.data,
            eventProps.topics,
          );

          return {
            args,
            hasArgumentNames: false,
            eventFragment: newEvent,
            signature: newEvent.format(FormatTypes.sighash),
            fullSignature: newEvent.format(FormatTypes.full),
          };
        } catch (err) {
          error = err;
          // We expect this to fail for some combinations. It's only a problem if it fails for all combinations.
        }
      }

      if (error) {
        throw error;
      }
    }
  }

  return logger.throwArgumentError('no matching event', 'topichash', topichash);
}

// @internal
export function omitNonNamedKeys<T extends object>(
  object: T,
): Record<string, string | BigNumber> | T {
  const obj: Record<string, string | BigNumber> = {};
  for (const [key, val] of Object.entries(object)) {
    if (isNaN(parseInt(key))) {
      obj[key] = val;
    }
  }
  return obj;
}

// @internal
export function doesContainNamedKeys<T extends object>(object: T): boolean {
  return Object.keys(object).some((key) => isNaN(parseInt(key)));
}

type ChangedParamType = Omit<ParamType, 'format' | '_isParamType'>;

function combinationsMap(
  inputs: ParamType[],
  indexedArgsCount: number,
  change: (param: ParamType) => ChangedParamType,
): ChangedParamType[][] {
  const combinations = generateCombinations(inputs.length, indexedArgsCount);
  return combinations.map((combination) =>
    inputs.map((input, idx) =>
      combination.includes(idx) ? change(input) : input,
    ),
  );
}
