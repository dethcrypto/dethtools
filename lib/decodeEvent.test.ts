import { EventFragment, FormatTypes, Interface } from '@ethersproject/abi'
import { expect } from 'earljs'

import { attachIndexedToJson, decodeEvent, doesContainNamedKeys, EventProps, filterNonNamedKeys } from './decodeEvent'
import { parseAbi } from './parseAbi'

describe(attachIndexedToJson.name, () => {
  it('attaches indexed modif. basing on topic count to json abi', () => {
    const iface = parseAbi(`
      'constructor(string symbol, string name)',
      'function transferFrom(address from, address to, uint256 amount)',
      'function getUser(uint256 id) view returns (tuple(string name, address addr) user)',
      'event Transfer(address indexed src, address indexed dst, uint256 amount)',`) as Interface

    expect(iface).toBeA(Interface)

    const indexedIface = attachIndexedToJson(iface.format(FormatTypes.json), 2)

    expect(indexedIface).toBeA(Interface)

    expect(Object.entries(indexedIface.events)[0][1].inputs[0].indexed).toEqual(true)
    expect(Object.entries(indexedIface.events)[0][1].inputs[1].indexed).toEqual(true)
    expect(Object.entries(indexedIface.events)[0][1].inputs[2].indexed).toBeFalsy()
  })
})

describe(filterNonNamedKeys.name, () => {
  it('filters non named keys e.g 0, 1 correctly', () => {
    expect(filterNonNamedKeys({ '0': 'hello', 1: 'world!', dethTools: 'is', awesome: '8)' })).toEqual({
      dethTools: 'is',
      awesome: '8)',
    })
  })
})

describe(doesContainNamedKeys.name, () => {
  it('checks if object contains named keys', () => {
    expect(doesContainNamedKeys({ '0': 'hello', 1: 'world!', dethTools: 'is', awesome: '8)' })).toEqual(true)
    expect(doesContainNamedKeys({ '0': 'hello', 1: 'world!' })).toEqual(false)
  })
})

describe(decodeEvent.name, () => {
  it('handles expected case', () => {
    const iface = parseAbi(`
      'constructor(string symbol, string name)',
      'function transferFrom(address from, address to, uint256 amount)',
      'function getUser(uint256 id) view returns (tuple(string name, address addr) user)',
      'event Transfer(address indexed src, address indexed dst, uint256 amount)',
      `) as Interface
    const eventProps: EventProps = {
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      topics: [
        '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
        '0x0000000000000000000000008ba1f109551bd432803012645ac136ddd64dba72',
        '0x000000000000000000000000ab7c8803962c0f2f5bbbe3fa8bf41cd82aa1923c',
      ],
    }
    const decodeEventResult = decodeEvent(iface, eventProps)

    expect(decodeEventResult).toBeDefined()

    const { decodedTopics, eventFragment, isNamedResult } = decodeEventResult!

    expect(decodedTopics.src).toEqual('0x8ba1f109551bD432803012645Ac136ddd64DBA72')
    expect(decodedTopics.dst).toEqual('0xaB7C8803962c0f2F5BBBe3FA8bf41cd82AA1923C')
    expect(decodedTopics.amount._hex).toEqual('0x0de0b6b3a7640000')
    expect(eventFragment[1]).toBeA(EventFragment)
    expect(isNamedResult).toEqual(true)
  })

  it('handles case with too few topics', () => {
    const iface = parseAbi(`
      'constructor(string symbol, string name)',
      'function transferFrom(address from, address to, uint256 amount)',
      'function getUser(uint256 id) view returns (tuple(string name, address addr) user)',
      'event Transfer(address indexed src, address indexed dst, uint256 amount)',
      `) as Interface
    const eventProps: EventProps = {
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      topics: [
        '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
        '0x0000000000000000000000008ba1f109551bd432803012645ac136ddd64dba72',
      ],
    }
    const decodeEventResult = decodeEvent(iface, eventProps)

    expect(decodeEventResult).toMatchSnapshot()
  })

  it('handles non-named params case', () => {
    const iface = parseAbi(`
      'constructor(string symbol, string name)',
      'function transferFrom(address from, address to, uint256 amount)',
      'function getUser(uint256 id) view returns (tuple(string name, address addr) user)',
      'event Transfer(address indexed, address indexed, uint256)',
      `) as Interface
    const eventProps: EventProps = {
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      topics: [
        '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
        '0x0000000000000000000000008ba1f109551bd432803012645ac136ddd64dba72',
        '0x000000000000000000000000ab7c8803962c0f2f5bbbe3fa8bf41cd82aa1923c',
      ],
    }
    const decodeEventResult = decodeEvent(iface, eventProps)

    const { isNamedResult } = decodeEventResult!

    expect(isNamedResult).toEqual(false)

    expect(decodeEventResult).toMatchSnapshot()
  })

  it('handles case where abi does not contain events', () => {
    const iface = parseAbi(`
      'constructor(string symbol, string name)',
      'function transferFrom(address from, address to, uint256 amount)',
      'function getUser(uint256 id) view returns (tuple(string name, address addr) user)',
      `) as Interface
    const eventProps: EventProps = {
      data: '0x0000000000000000000000000000000000000000000000000de0b6b3a7640000',
      topics: [
        '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
        '0x0000000000000000000000008ba1f109551bd432803012645ac136ddd64dba72',
        '0x000000000000000000000000ab7c8803962c0f2f5bbbe3fa8bf41cd82aa1923c',
      ],
    }
    const decodeEventResult = decodeEvent(iface, eventProps)

    expect(decodeEventResult).toMatchSnapshot()
  })
})
