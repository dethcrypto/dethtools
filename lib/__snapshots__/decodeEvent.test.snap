// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decodeEvent handles non-named params case 1`] = `
Object {
  "decodedTopics": Array [
    "0x8ba1f109551bD432803012645Ac136ddd64DBA72",
    "0xaB7C8803962c0f2F5BBBe3FA8bf41cd82AA1923C",
    Object {
      "hex": "0x0de0b6b3a7640000",
      "type": "BigNumber",
    },
  ],
  "eventFragment": Array [
    "Transfer(address,address,uint256)",
    EventFragment {
      "_isFragment": true,
      "anonymous": false,
      "inputs": Array [
        ParamType {
          "_isParamType": true,
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": true,
          "name": null,
          "type": "address",
        },
        ParamType {
          "_isParamType": true,
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": true,
          "name": null,
          "type": "address",
        },
        ParamType {
          "_isParamType": true,
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": null,
          "name": null,
          "type": "uint256",
        },
      ],
      "name": "Transfer",
      "type": "event",
    },
  ],
  "isNamedResult": false,
}
`;
